### DISCUSSION TITLE
feature importance ignoring pruned trials?

### DISCUSSION THREAD

**Discussion by mreitschuster** [optuna/optuna#3703] [Q&A] [answered]
Hi all!

Is my understanding correct that for parameter importance evaluation, e.g.
```python
optuna.visualization.plot_optimization_history(study1).show(renderer="svg")
optuna.visualization.plot_intermediate_values(study1).show(renderer="svg")
```

the pruned trials are not considered? Is there a way to consider them as a fixed value (e.g. min of all other one or manually provided one)?
Of course one can see the feature importances as "given all non-pruned trials", but that risks a wrong, intuitive interpretation of them. Imagine a study: 
```
feature 1 == A  -> 15x PRUNED
feature 1 == B  -> 15x COMPLETE
feature 1 == B  & feature 2=X   -> 7x VALUE 1   
feature 1 == B  & feature 2=Y   -> 8x VALUE 1.1
```   
My understanding would be that feature 1 is very important, while i dont think feature 2 is important. But what I think I get from optuna is a "feature 1 is not important at all".

I tried to get the complementary analysis "what features lead to pruning", which in itself is not great as the pruning depends on the order in which the trials are executed. But I run into an error anyway: 

```python
optuna.visualization.plot_param_importances(
    study1, target=lambda t: (t.state==optuna.trial.TrialState['PRUNED']), target_name="pruned"
).show(renderer="svg")
RuntimeError: Encountered zero total variance in all trees.
```
I then tried playing around with a wrapper function that gives trial.value when available and 0 when not. But it seem the check to exclude all non-omplete trial states is hardcoded in optuna/importance/_fanova/_evaluator.py line 92-97. So i would need to have the wrapper overwrite the state. But that seems not to be possible
```python
study1.trials[15].state=optuna.trial.TrialState['COMPLETE']
study1.trials[15].state
Out[52]: TrialState.PRUNED
```

```python
study_new.tell(trial=15, values=0, state=optuna.trial.TrialState['COMPLETE'])
RuntimeError: Trial#15 has already finished and can not be updated.
```

So I see only 3 options: **rewrite _evaluator.py** or **modify the database entries** - both pieces that will trigger an avalanche of other issues, so I want to avoid. or **rerun with NoOp pruner**.

How would you deal with this - getting reliable feature importance analysis that includes the information that a trial is pruned?

**Comment by nzw0301** [comment, marked as answer]
Hi, welcome Optuna's github discussions.


> the pruned trials are not considered?

Yes, your understanding is correct.

> Is there a way to consider them as a fixed value (e.g. min of all other one or manually provided one)?

Yes. A possible way is to create a new study from the study and change the state from `PRUNED` to `COMPLETE`, as follows:



```python
import optuna
study = optuna.create_study()

trial = study.ask()
trial.suggest_int("x", 0, 100)
trial.suggest_int("y", 0, 100)
study.tell(trial, 1.0, optuna.trial.TrialState.COMPLETE)
trial = study.ask()
trial.suggest_int("x", 0, 100)
trial.suggest_int("y", 0, 100)
study.tell(trial, 0.8, optuna.trial.TrialState.COMPLETE)

# the following trial is ignored
trial = study.ask()
trial.suggest_int("x", 0, 100)
trial.suggest_int("y", 0, 100)
study.tell(trial, state=optuna.trial.TrialState.PRUNED)

optuna.visualization.plot_param_importances(study)

trials = []
for t in study.get_trials():
    if t.state == optuna.trial.TrialState.PRUNED:
        t = optuna.trial.create_trial(
            state=optuna.trial.TrialState.COMPLETE,
            value=0.5, # maybe min/max(intermediate_values) or intermediate_values[-1]
            params=t.params, distributions=t.distributions,
            user_attrs=t.user_attrs, system_attrs=t.system_attrs,
            intermediate_values=t.intermediate_values)
        trials.append(t)
    elif t.state == optuna.trial.TrialState.COMPLETE:
        trials.append(t)
    else:
        # write something else for another trial state
        assert False

study_for_importance_param = optuna.create_study()
study_for_importance_param.add_trials(trials)
optuna.visualization.plot_param_importances(study_for_importance_param)
```

**Reply by mreitschuster** [reply]
Thank you for the quick answer! Works like a charm.

I have packed it into a function for my own use.
```python
import optuna
def set_PRUNED_to_value(study, value):
    trials = []
    for t in study.get_trials():
        if t.state == optuna.trial.TrialState.PRUNED:
            t = optuna.trial.create_trial(
                state=optuna.trial.TrialState.COMPLETE,
                value=value,
                params=t.params, 
                distributions=t.distributions,
                user_attrs=t.user_attrs, 
                system_attrs=t.system_attrs,
                intermediate_values=t.intermediate_values)
            trials.append(t)
        elif t.state == optuna.trial.TrialState.COMPLETE:
            trials.append(t)
        else:
            pass
    new_study=optuna.create_study()
    new_study.add_trials(trials)
    return new_study
```

Maybe you want to consider either changing default behavior (use pruned trials for feature importance) or make it explicit to the user, that feature that lead to pruning are not considered important. I believe that otherwise users will misinterpret the results easily.

**Reply by nzw0301** [reply]
Thank you for your confirmation and feedback!

 Indeed, [pruned trial has an objective value when it has intermediate values](https://github.com/optuna/optuna/issues/3542), so the following sentence on [the importance evaluator page](https://optuna.readthedocs.io/en/latest/reference/generated/optuna.importance.FanovaImportanceEvaluator.html):

> Given a study, fANOVA fits a random forest regression model that predicts the objective value given a parameter configuration. 

sounds inaccurate as you mentioned. So it would be great to clarify the current behaviour to avoid confusion by users at least. I suppose considering pruned trials for calculation of importance scores needs discussion, please create a feature request issue if you are interested in the feature.

Best regards,

---

PS: I've created a PR to clarify the docs. see https://github.com/optuna/optuna/pull/3707.

**Reply by mreitschuster** [reply]
Thanks. I have opened [a feature request](https://github.com/optuna/optuna/issues/3709)

**Reply by nzw0301** [reply]
Great. Thanks!
